pushsync/pusher.go

```diff
@@ -116,119 +114,132 @@ func (p *Pusher) sync() {
+				// increment synced count for the tag if exists
+				tag := item.tag
+				if tag != nil {
+					syncedTags[tag.Uid] = syncedTags[tag.Uid] + 1
+					item.span.Finish()
+				}
+				delete(p.pushed, syncedAddrs[i].Hex())
 			}
-			for _, addr := range synced {
-				// set chunk status to synced, insert to db GC index
-				if err := p.store.Set(context.Background(), chunk.ModeSetSync, addr); err != nil {
-					log.Warn("error setting chunk to synced", "addr", addr, "err", err)
-					continue
+			// iterate over tags in this batch
+			for uid, n := range syncedTags {
+				tag, _ := p.tags.Get(uid)
+				tag.IncN(chunk.StateSynced, n)
+				if tag.Done(chunk.StateSynced) {
+					log.Info("closing root span for tag", "taguid", tag.Uid, "tagname", tag.Name)
+					tag.FinishRootSpan()
 				}
-				delete(p.pushed, addr.Hex())
 			}
 			// reset synced list
-			synced = nil
+			syncedAddrs = nil
+			syncedItems = nil
 
 			// we don't want to record the first iteration
 			if chunksInBatch != -1 {
 				// this measurement is not a timer, but we want a histogram, so it fits the data structure
 				metrics.GetOrRegisterResettingTimer("pusher.subscribe-push.chunks-in-batch.hist", nil).Update(time.Duration(chunksInBatch))
-
 				metrics.GetOrRegisterResettingTimer("pusher.subscribe-push.chunks-in-batch.time", nil).UpdateSince(batchStartTime)
 				metrics.GetOrRegisterCounter("pusher.subscribe-push.chunks-in-batch", nil).Inc(int64(chunksInBatch))
 			}
 			chunksInBatch = 0
 			batchStartTime = time.Now()
 
 			// and start iterating on Push index from the beginning
-			ctx, cancel = context.WithCancel(context.Background())
-			chunks, stop = p.store.SubscribePush(ctx)
+			chunks, unsubscribe = p.store.SubscribePush(ctx)
 			// reset timer to go off after retryInterval
 			timer.Reset(retryInterval)
 
 		// handle incoming chunks
 		case ch, more := <-chunks:
-			func() {
-				chunksInBatch++
-				// if no more, set to nil and wait for timer
-				if !more {
-					chunks = nil
-					return
-				}
+			// if no more, set to nil, reset timer to 0 to finalise batch immediately
+			if !more {
+				chunks = nil
+				timer.Reset(0)
+				break
+			}
 
-				metrics.GetOrRegisterCounter("pusher.send-chunk", nil).Inc(1)
-				// if no need to sync this chunk then continue
-				if !p.needToSync(ch) {
-					return
-				}
+			chunksInBatch++
+			metrics.GetOrRegisterCounter("pusher.send-chunk", nil).Inc(1)
+			// if no need to sync this chunk then continue
+			if !p.needToSync(ch) {
+				break
+			}
 
-				metrics.GetOrRegisterCounter("pusher.send-chunk.send-to-sync", nil).Inc(1)
-				// send the chunk and ignore the error
-				go func(ch chunk.Chunk) {
-					if err := p.sendChunkMsg(ch); err != nil {
-						log.Error("error sending chunk", "addr", ch.Address(), "err", err)
-					}
-				}(ch)
-			}()
+			metrics.GetOrRegisterCounter("pusher.send-chunk.send-to-sync", nil).Inc(1)
+			// send the chunk and ignore the error
+			// go func(ch chunk.Chunk) {
```

Comment:
Please remove commented code

Reply:
> @zelig since I'm going on vacation I'm approving this with the prospect of this being merged soon and because I don't want to block this (it is also in a good shape).
> 

thanks

> There are still unresolved and unaddressed comments from the previous round of reviews, I'd be happy if you could at least comment on why you think they are irrelevant (since they were not addressed, I am assuming they were considered irrelevant).
> 

I tried to address each one of them, anything in particular 

> In specific I don't like the fact that the other unresolved dangling PR about the fetcher logic change is piggybacked into this PR. @nonsense did you have anything to do with this? did you two reach some off-band agreement about this topic that it has been decided to pull this into this PR?

good spotting, maybe i just omit that change

> 
> Also, I don't like the interface naming convention freewheeling habits that we have. I have never seen a `PubSub` interface that defined anything more than two methods `Publish` and `Subscribe` and I have a hard time understanding why we need to choose counter intuitive names for everything

cmon, shall i call it `TopicMessenger` or what? this PubSub interface has other method names